apis/r/R/write_seurat.R

```diff
@@ -207,6 +208,210 @@ write_soma.Assay <- function(
+#'   measurement level; indicates the Seurat assay version.
+#'   Set to \dQuote{\code{v5}}
+#'  \item \dQuote{\code{soma_ecosystem_seurat_v5_default_layers}}: written at
+#'   the \dQuote{\code{X}} group level; indicates the
+#'   \link[SeuratObject:DefaultLayer]{default layers}
+#'  \item \dQuote{\code{soma_ecosystem_seurat_v5_ragged}}: written at the
+#'   \dQuote{\code{X/<layer>}} array level; with a value of
+#'   \dQuote{\code{ragged}}, indicates whether or not the layer is ragged
+#'  \item \dQuote{\code{soma_r_type_hint}}: written at the
+#'   \dQuote{\code{X/<layer>}} array level; indicates the \R class and
+#'   defining package (for S4 classes) of the original layer
+#' }
+#'
+#' @method write_soma Assay5
+#' @export
+#'
+write_soma.Assay5 <- function(
+  x,
+  uri = NULL,
+  soma_parent,
+  ...,
+  ingest_mode = 'write',
+  platform_config = NULL,
+  tiledbsoma_ctx = NULL,
+  relative = TRUE
+) {
+  check_package('SeuratObject', version = '5.0.1')
+  stopifnot(
+    "'uri' must be a single character value" = is.null(uri) ||
+      is_scalar_character(uri),
+    "'soma_parent' must be a SOMACollection" = inherits(
+      x = soma_parent,
+      what = 'SOMACollectionBase'
+    ),
+    "'relative' must be a single logical value" = is_scalar_logical(relative)
+  )
+
+  # Find `shape` if and only if we're called from `write_soma.Seurat()`
+  parents <- unique(sys.parents())
+  idx <- which(vapply_lgl(
+    parents,
+    FUN = function(i) identical(sys.function(i), write_soma.Seurat)
+  ))
+  shape <- if (length(idx) == 1L) {
+    get("shape", envir = sys.frame(parents[idx]))
+  } else {
+    NULL
+  }
+  shape <- rev(shape)
+
+  # Create a proper URI
+  uri <- uri %||% gsub(pattern = '_$', replacement = '', x = SeuratObject::Key(x))
+  uri <- .check_soma_uri(
+    uri = uri,
+    soma_parent = soma_parent,
+    relative = relative
+  )
+
+  # Create the measurement
+  ms <- SOMAMeasurementCreate(
+    uri = uri,
+    ingest_mode = ingest_mode,
+    platform_config = platform_config,
+    tiledbsoma_ctx = tiledbsoma_ctx
+  )
+  ms$set_metadata(.assay_version_hint('v5'))
+  X <- if (!'X' %in% ms$names()) {
+    SOMACollectionCreate(
+      uri = file_path(ms$uri, 'X'),
+      ingest_mode = ingest_mode,
+      platform_config = platform_config,
+      tiledbsoma_ctx = tiledbsoma_ctx
+    )
+  } else if (isTRUE(relative)) {
+    SOMACollectionOpen(uri = file_path(ms$uri, 'X'), mode = 'WRITE')
+  } else {
+    ms$X
+  }
+  withCallingHandlers(
+    .register_soma_object(X, soma_parent = ms, key = 'X', relative = relative),
+    existingKeyWarning = .maybe_muffle
+  )
+  on.exit(X$close(), add = TRUE, after = FALSE)
+
+  X$set_metadata(.layer_hint(SeuratObject::DefaultLayer(x)))
+
+  # Pull presence matrices from the v5 assay
+  cmat <- methods::slot(x, name = 'cells')
+  fmat <- methods::slot(x, name = 'features')
+
+  # Write `X` matrices
+  for (lyr in SeuratObject::Layers(x)) {
```

Comment:
I think spelling out `lyr` as `layer` is a reasonable ask

also I think it's a `layer_name` not a `layer` -- given `key = lyr` below